### PR TITLE
Documentation: Add support for deprecated/experimental messages

### LINKED ISSUE #7900: "Deprecated" needs to be more visible in the documentation
What it looks like : 

bild

What it should look like : 

bild

I just wasted several hours asking around on help forums why I couldn't find the class as the documentation is misleading, and this deprecated isn't yet fresh in the mind of "beginner" helpers.
For anyone not familiar with Godot's ecosystem, "up to date as of 4.1" could as well mean "4.1 **onwards**"

### PR DESCRIPTION
**Pull request by dalexeev**
Add support for deprecated/experimental messages to core and GDScript. See:

* godotengine/godot-proposals#6559
* godotengine/godot-proposals#7622

```gdscript
## my_func desc.
## @deprecated: Use [method another] instead.
func my_func():
    pass
```

* *Bugsquad edit, fixes: https://github.com/godotengine/godot-docs/issues/7900*

* _Production edit: closes godotengine/godot-roadmap#71_

### CODE REVIEW

**Comment by YuriSizov** [comment]
This would be a breaking change for the XML format. We should find a way to avoid that.

**Comment by dalexeev** [comment]
We can use two attributes, bool and string, just like in `DocData` structures. But it's less compact. Or we can add compatibility code for the old boolean attributes, which will allow the old XMLs to be read, but on write they will be saved in the new format. It's funny if we deprecate the `is_deprecated` attribute.

**Comment by raulsntos** [comment]
I love this, we've been generating generic messages like `This method is deprecated` in the generated C# API and now we'll be able to provide something more informative.

I noticed there were some empty deprecation messages. Should we try to avoid empty messages or is it intended that some messages will be empty? We can still provide the generic message in C# in these cases, but I think it'd be better to provide a better message. I'm not suggesting to fill all the messages in this PR, it could be a follow-up.

I'm not sure how useful it is for experimental API though, it seems all the messages are empty in this PR. I also don't know what we would write here other than a generic `This API is experimental and may change in the future`. We also wouldn't be able to use these messages in C#, the `[Experimental]` attribute doesn't allow providing a message, but that shouldn't block this if it's still useful.

File: modules/gltf/doc_classes/GLTFDocument.xml [hunk 1]
```diff
@@ -9,7 +9,7 @@
 		GLTFDocument can be extended with arbitrary functionality by extending the [GLTFDocumentExtension] class and registering it with GLTFDocument via [method register_gltf_document_extension]. This allows for custom data to be imported and exported.
 	</description>
 	<tutorials>
-		<link title="glTF 'What the duck?' guide">https://www.khronos.org/files/gltf20-reference-guide.pdf</link>
+		<link title="glTF &apos;What the duck?&apos; guide">https://www.khronos.org/files/gltf20-reference-guide.pdf</link>
```
Comment (raulsntos) [inline modules/gltf/doc_classes/GLTFDocument.xml:12]: This change seems unrelated.

File: doc/classes/SkeletonIK3D.xml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8" ?>
-<class name="SkeletonIK3D" inherits="Node" is_deprecated="true" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+<class name="SkeletonIK3D" inherits="Node" deprecated="" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
```
Comment (raulsntos) [inline doc/classes/SkeletonIK3D.xml:2]: I think there was someone asking recently why this class was deprecated, worried that it will be removed without an alternative. Maybe we could add a message here that clarifies that a replacement will be provided in the future?

**Comment by dalexeev** [comment]
> I noticed there were some empty deprecation messages. Should we try to avoid empty messages or is it intended that some messages will be empty?

I think so, but I've left the messages empty for now. We should also think about default messages. Now I've added a neutral "This ... may be changed or removed in future versions.", but this should probably be different for Deprecated and Experimental and for engine and user code.

> I'm not sure how useful it is for experimental API though, it seems all the messages are empty in this PR. I also don't know what we would write here other than a generic `This API is experimental and may change in the future`.

Probably yes, although I think that sometimes we can indicate problems when using the experimental API:

File: modules/gltf/doc_classes/GLTFDocument.xml (hunk 1 above)
Comment (dalexeev) [reply modules/gltf/doc_classes/GLTFDocument.xml:12]: I added missing quote escaping in `DocTools`.

File: doc/class.xsd [hunk 3]
```diff
@@ -273,8 +289,13 @@
 			</xs:sequence>
 			<xs:attribute type="xs:string" name="name" />
 			<xs:attribute type="xs:string" name="inherits" />
+			<!-- deprecated -->
+			<xs:attribute type="xs:float" name="version" use="optional" />
```
Comment (dalexeev) [inline doc/class.xsd:293]: * See #79092.

File: doc/tools/make_rst.py [hunk 4]
```diff
@@ -1443,6 +1487,28 @@ def make_getter_signature(class_def: ClassDef, property_def: PropertyDef, state:
...
+def make_deprecated_experimental(item: DefinitionBase, state: State) -> str:
+    result = ""
+
+    if item.deprecated is not None:
+        deprecated_prefix = translate("Deprecated:")
+        if item.deprecated.strip() == "":
+            default_message = translate(f"This {item.definition_name} may be changed or removed in future versions.")
+            result += f"**{deprecated_prefix}** {default_message}\n\n"
+        else:
+            result += f"**{deprecated_prefix}** {format_text_block(item.deprecated.strip(), item, state)}\n\n"
```
Comment (dalexeev) [inline doc/tools/make_rst.py:1499]: Is there a way to use color in RST? This will probably require some changes to the site so that we can use:

```rst
:red:`Deprecated:"` Message.
```

or

```rst
.. deprecated:: Message.
```

Now it looks like this:



I only added this to the descriptions, not to the overviews.

Comment (Calinou) [reply doc/tools/make_rst.py:1499]: RST doesn't support arbitrary colors (aside of raw HTML tags).

Comment (dalexeev) [reply doc/tools/make_rst.py:1499]: [How to use color in text with ReStructured Text (rst2html.py) or how to insert HTML tags without blank lines?](https://stackoverflow.com/questions/4669689/how-to-use-color-in-text-with-restructured-text-rst2html-py-or-how-to-insert-h)

**Review by AThousandShips — APPROVED** [review]
LGTM style and code wise

**Review by Calinou — COMMENTED** [review]
Tested locally (rebased on top of `master` f317cc713aa4dbcee2efa10db764473a56680be7), it mostly works as expected.

I can't get the deprecated/experimental reason messages to show up though:

```gdscript
extends Node2D

## Description.
##@deprecated: Test reason
var var1

##@experimental: Test reason 2
var var2

# Test comment (not a docblock).
##@deprecated: Test reason 3
func one() -> void:
	pass


##@experimental: Test reason 4
##Description.
func two() -> void:
	pass
```

image

**Comment by dalexeev** [comment]
@Calinou Looks like you were testing an older version:

File: doc/classes/@GlobalScope.xml [hunk 5]
```diff
@@ -2904,9 +2903,8 @@
 		<constant name="PROPERTY_HINT_GLOBAL_SAVE_FILE" value="28" enum="PropertyHint">
 			Hints that a [String] property is a path to a file. Editing it will show a file dialog for picking the path for the file to be saved at. The dialog has access to the entire filesystem. The hint string can be a set of filters with wildcards like [code]"*.png,*.jpg"[/code]. See also [member FileDialog.filters].
 		</constant>
-		<constant name="PROPERTY_HINT_INT_IS_OBJECTID" value="29" enum="PropertyHint" is_deprecated="true">
+		<constant name="PROPERTY_HINT_INT_IS_OBJECTID" value="29" enum="PropertyHint" deprecated="This hint is not used anywhere and will be removed in the future.">
 			Hints that an [int] property is an object ID.
```
Comment (akien-mga) [inline doc/classes/@GlobalScope.xml:2907]: The description should probably be removed as the hint doesn't seem to be implemented at all.

File: doc/classes/@GlobalScope.xml [hunk 6]
```diff
@@ -2977,9 +2975,8 @@
 		<constant name="PROPERTY_USAGE_UPDATE_ALL_IF_MODIFIED" value="16384" enum="PropertyUsageFlags" is_bitfield="true">
 			If this property is modified, all inspector fields will be refreshed.
 		</constant>
-		<constant name="PROPERTY_USAGE_SCRIPT_DEFAULT_VALUE" value="32768" enum="PropertyUsageFlags" is_bitfield="true" is_deprecated="true">
+		<constant name="PROPERTY_USAGE_SCRIPT_DEFAULT_VALUE" value="32768" enum="PropertyUsageFlags" is_bitfield="true" deprecated="This hint is not used anywhere and will be removed in the future.">
 			Signifies a default value from a placeholder script instance.
```
Comment (akien-mga) [inline doc/classes/@GlobalScope.xml:2979]: Likewise here.

File: doc/classes/@GlobalScope.xml [hunk 7]
```diff
@@ -3008,9 +3005,8 @@
 		<constant name="PROPERTY_USAGE_KEYING_INCREMENTS" value="16777216" enum="PropertyUsageFlags" is_bitfield="true">
 			Inserting an animation key frame of this property will automatically increment the value, allowing to easily keyframe multiple values in a row.
 		</constant>
-		<constant name="PROPERTY_USAGE_DEFERRED_SET_RESOURCE" value="33554432" enum="PropertyUsageFlags" is_bitfield="true" is_deprecated="true">
+		<constant name="PROPERTY_USAGE_DEFERRED_SET_RESOURCE" value="33554432" enum="PropertyUsageFlags" is_bitfield="true" deprecated="This hint is not used anywhere and will be removed in the future.">
 			When loading, the resource for this property can be set at the end of loading.
```
Comment (akien-mga) [inline doc/classes/@GlobalScope.xml:3009]: And here.

File: doc/classes/AnimationPlayer.xml [hunk 8]
```diff
@@ -44,10 +44,10 @@
 				Returns the blend time (in seconds) between two animations, referenced by their keys.
 			</description>
 		</method>
-		<method name="get_method_call_mode" qualifiers="const" is_deprecated="true">
+		<method name="get_method_call_mode" qualifiers="const" deprecated="">
 			<return type="int" enum="AnimationPlayer.AnimationMethodCallMode" />
 			<description>
-				For backward compatibility. See [enum AnimationMixer.AnimationCallbackModeMethod].
+				See [enum AnimationMixer.AnimationCallbackModeMethod].
```
Comment (akien-mga) [inline doc/classes/AnimationPlayer.xml:50]: Should this be the `deprecated=""` message instead, as "Use ... instead."?

Same for the others.

File: doc/classes/InputEventJoypadButton.xml [hunk 9]
```diff
@@ -16,9 +16,8 @@
 		<member name="pressed" type="bool" setter="set_pressed" getter="is_pressed" default="false">
 			If [code]true[/code], the button's state is pressed. If [code]false[/code], the button's state is released.
 		</member>
-		<member name="pressure" type="float" setter="set_pressure" getter="get_pressure" default="0.0" is_deprecated="true">
+		<member name="pressure" type="float" setter="set_pressure" getter="get_pressure" default="0.0" deprecated="This property is never set by the engine and is always [code]0[/code].">
 			Represents the pressure the user puts on a pressure-sensitive button.
```
Comment (akien-mga) [inline doc/classes/InputEventJoypadButton.xml:20]: I think this was discussed recently in a PR by @Mickeon, but I still think we shouldn't be documenting what a property _ought to do_ if it doesn't do it. If this is never set, that's everything the user needs to know, not what some potential thirdparties may decide to represent with this unused property.

File: doc/classes/MultiMesh.xml [hunk 10]
```diff
@@ -90,10 +90,10 @@
 	<members>
 		<member name="buffer" type="PackedFloat32Array" setter="set_buffer" getter="get_buffer" default="PackedFloat32Array()">
 		</member>
-		<member name="color_array" type="PackedColorArray" setter="_set_color_array" getter="_get_color_array" is_deprecated="true">
+		<member name="color_array" type="PackedColorArray" setter="_set_color_array" getter="_get_color_array" deprecated="">
 			See [method set_instance_color].
```
Comment (akien-mga) [inline doc/classes/MultiMesh.xml:94]: Could also be the deprecation message.

This brings the question of how should deprecated properties and methods be documented in general though, I'll bring this up in the main PR comments.

File: doc/classes/SkeletonIK3D.xml (hunk 2 above)
Comment (akien-mga) [reply doc/classes/SkeletonIK3D.xml:2]: CC @godotengine/animation

**Review by akien-mga — APPROVED** [review]
Looks great to me overall!

**Comment by akien-mga** [comment]
We should probably find some consensus on how to documentation deprecated and experimental methods/properties, and apply it throughout the codebase (cc @godotengine/documentation).

Notably, how to document a method/property when it's deprecated because:
- It doesn't do anything and we're just keeping it to avoid breaking API compat
- It still does something but was replaced by a new method/property which should be referenced
- It still does something but is no longer supported, and no alternative is given

We can now put information in the deprecated message, but there's still the main description to deal with. What should it contain? Should it be kept empty if all the relevant info is in the deprecation message? If so, should we change `doc/tools/doc_status.py` so it considers an empty description of a method/property with a deprecated message as fully documented?

File: doc/classes/InputEventJoypadButton.xml (hunk 9 above)
Comment (Mickeon) [reply doc/classes/InputEventJoypadButton.xml:20]: Because the **Deprecated** message comes first in the PR now, this whole description to me is off. I don't think saying "Represents the pressure[...]" makes sense anymore because it feels a bit... contradictory? At best it should be "Should represent the pressure[...]". At worst, it should not exist at all.

File: doc/classes/TileMap.xml [hunk 11]
```diff
@@ -76,11 +76,10 @@
 				Clears cells that do not exist in the tileset.
 			</description>
 		</method>
-		<method name="force_update" is_deprecated="true">
+		<method name="force_update" deprecated="See [method notify_runtime_tile_data_update] and [method update_internals].">
 			<return type="void" />
 			<param index="0" name="layer" type="int" default="-1" />
 			<description>
-				[i]Deprecated.[/i] See [method notify_runtime_tile_data_update] and [method update_internals].
```
Comment (Mickeon) [inline doc/classes/TileMap.xml:83]: I'm a bit peeved off by these kind of descriptions.

Obviously, `force_update` still works, so it needs to have a _somewhat_ meaningful description still, albeit extremely succinct, so that users are even more encouraged not to use it. Yet they would still know what it does if they encounter it in old code.

Deprecating without the "what" and "why" is is a bit malicious in my opinion.

**Comment by Mickeon** [comment]
I was taking a look. In my opinion, to make it extremely brief:
- The deprecated/experimental message should explain "Why"
- The description should explain "what". Very, **very** briefly.
    - I can't think of exceptions. Not saying what _something_ is/was feels wrong to me.


<sub>Also these messages are not translatable now, are they?</sub>

**Comment by dalexeev** [comment]
> Also these messages are not translatable now, are they?

Yes, after merging this PR, some changes are needed in [extract_classes.py](https://github.com/godotengine/godot-editor-l10n/blob/main/scripts/extract_classes.py).

As for the docs, I agree with you, but I think that in any case we will need a follow-up PR for adjustments. However, I'm ready to apply suggestions that we are already confident about.

**Comment by Mickeon** [comment]
I don't mind scouting across the class reference a second time as some form of standardization. I got nothing better to do...

**Comment by akien-mga** [comment]
Thanks!
